analyzer/consensus/consensus.go

```diff
@@ -122,25 +107,33 @@ func (m *processor) PreWork(ctx context.Context) error {
 	return nil
 }
 
-func (m *processor) isGenesisProcessed(ctx context.Context, chainContext string) (bool, error) {
-	var processed bool
-	if err := m.target.QueryRow(
-		ctx,
-		queries.IsGenesisProcessed,
-		chainContext,
-	).Scan(&processed); err != nil {
-		return false, err
+// Implements block.BlockProcessor interface. Downloads and processes the genesis document
+// that immediately precedes block `lastFastSyncHeight`+1, i.e. the first slow-sync block
+// we're about to process.
+// If that block is the first block of a chain (cobalt, damask, etc), we download the chain's
+// original genesis document. Otherwise, we download the genesis-formatted state at `lastFastSyncHeight`.
+func (m *processor) FinalizeFastSync(ctx context.Context, lastFastSyncHeight int64) error {
+	firstSlowSyncHeight := lastFastSyncHeight + 1
+	r, err := m.history.RecordForHeight(firstSlowSyncHeight)
+	if err != nil {
+		return fmt.Errorf("no history record for first slow-sync height %d: %w", firstSlowSyncHeight, err)
+	}
+	var genesisDoc *genesis.Document
+	if r.GenesisHeight == firstSlowSyncHeight {
+		m.logger.Info("fetching genesis document before starting with the first block of a chain", "chain_context", r.ChainContext, "genesis_height", r.GenesisHeight)
+		genesisDoc, err = m.source.GenesisDocument(ctx, r.ChainContext)
```

Comment:
Does this return a different result to what `StateToGenesis` would return if called at the genesis height?